modules/tensor_mechanics/src/materials/ComputePlasticTrussResultants.C

```diff
@@ -0,0 +1,158 @@
+ComputePlasticTrussResultants::validParams()
+{
+  InputParameters params = ComputeTrussResultants::validParams();
+  params.addClassDescription(
+      "Computes the stress and strain for a truss element with plastic behavior defined by either "
+      "linear hardening or a user-defined hardening function.");
+  params.addRequiredParam<Real>("yield_stress",
+                                "Yield stress after which plastic strain starts accumulating");
+  params.addParam<Real>("hardening_constant", 0.0, "Hardening slope");
+  params.addParam<FunctionName>("hardening_function",
+                                "Engineering stress as a function of plastic strain");
+  params.addRequiredCoupledVar(
+      "area",
+      "Cross-section area of the truss. Can be supplied as either a number or a variable name.");
+  params.addParam<Real>(
+      "absolute_tolerance", 1e-10, "Absolute convergence tolerance for Newton iteration");
+  params.addParam<Real>(
+      "relative_tolerance", 1e-8, "Relative convergence tolerance for Newton iteration");
+  return params;
+}
+
+ComputePlasticTrussResultants::ComputePlasticTrussResultants(const InputParameters & parameters)
+  : ComputeTrussResultants(parameters),
+    _yield_stress(getParam<Real>("yield_stress")), // Read from input file
+    _hardening_constant(getParam<Real>("hardening_constant")),
+    _hardening_function(isParamValid("hardening_function") ? &getFunction("hardening_function")
+                                                           : NULL),
+    _area(coupledValue("area")),
+    _material_stiffness(getMaterialPropertyByName<Real>("material_stiffness")),
+    _total_disp_strain(getMaterialProperty<RealVectorValue>("total_disp_strain")),
+    _total_disp_strain_old(getMaterialPropertyOld<RealVectorValue>("total_disp_strain")),
+
+    _absolute_tolerance(parameters.get<Real>("absolute_tolerance")),
+    _relative_tolerance(parameters.get<Real>("relative_tolerance")),
+
+    _plastic_strain(declareProperty<Real>("plastic_strain")),
+    _plastic_strain_old(getMaterialPropertyOld<Real>("plastic_strain")),
+
+    _axial_stress(declareProperty<Real>("axial_stress")),
+    _axial_stress_old(getMaterialPropertyOld<Real>("axial_stress")),
+    _force(declareProperty<Real>("forces")),
+    _force_old(getMaterialPropertyOld<Real>("forces")),
+    _hardening_variable(declareProperty<Real>("hardening_variable")),
+    _hardening_variable_old(getMaterialPropertyOld<Real>("hardening_variable")),
+    _max_its(1000)
+{
+  if (!parameters.isParamSetByUser("hardening_constant") && !isParamValid("hardening_function"))
+    mooseError("ComputePlasticTrussResultants: Either hardening_constant or hardening_function "
+               "must be defined");
+
+  if (parameters.isParamSetByUser("hardening_constant") && isParamValid("hardening_function"))
+    mooseError("ComputePlasticTrussResultants: Only the hardening_constant or only the "
+               "hardening_function can be defined but not both");
```

Comment:
I suggest just providing a `hardening_function` parameter. Users can supply simple numbers for that parameter as well, no need to complicate it.

Reply:
never mind, I see the calculation for the two is different, then just soimplify to

```suggestion
  if (parameters.isParamSetByUser("hardening_constant") == isParamValid("hardening_function"))
    mooseError("ComputePlasticTrussResultants: Either hardening_constant or hardening_function "
               "must be defined");
```

Reply:
I modified it now.